crates/evm-ethereum/src/execute.rs

```diff
@@ -0,0 +1,826 @@
+
+    fn executor<DB>(&self, db: DB) -> Self::Executor<DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        self.eth_executor(db)
+    }
+
+    fn batch_executor<DB>(&self, db: DB) -> Self::BatchExecutor<DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        let executor = self.eth_executor(db);
+        EthBatchExecutor {
+            executor,
+            batch_record: BlockBatchRecord::new(self.prune_modes.clone()),
+            stats: BlockExecutorStats::default(),
+        }
+    }
+}
+
+/// Helper container type for EVM with chain spec.
+#[derive(Debug, Clone)]
+struct EthEvmExecutor<EvmConfig> {
+    /// The chainspec
+    chain_spec: Arc<ChainSpec>,
+    /// How to create an EVM.
+    evm_config: EvmConfig,
+}
+
+impl<EvmConfig> EthEvmExecutor<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = ()>,
+{
+    /// Executes the transactions in the block and returns the receipts.
+    ///
+    /// This applies the pre-execution changes, and executes the transactions.
+    ///
+    /// # Note
+    ///
+    /// It does __not__ apply post-execution changes.
+    fn execute_pre_and_transactions<Ext, DB>(
+        &mut self,
+        block: &BlockWithSenders,
+        mut evm: Evm<'_, Ext, &mut State<DB>>,
+    ) -> Result<(Vec<Receipt>, u64), BlockExecutionError>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        // apply pre execution changes
+        apply_beacon_root_contract_call(
+            &self.chain_spec,
+            block.timestamp,
+            block.number,
+            block.parent_beacon_block_root,
+            &mut evm,
+        )?;
+
+        // execute transactions
+        let mut cumulative_gas_used = 0;
+        let mut receipts = Vec::with_capacity(block.body.len());
+        for (sender, transaction) in block.transactions_with_sender() {
+            // The sum of the transaction’s gas limit, Tg, and the gas utilized in this block prior,
+            // must be no greater than the block’s gasLimit.
+            let block_available_gas = block.header.gas_limit - cumulative_gas_used;
+            if transaction.gas_limit() > block_available_gas {
+                return Err(BlockValidationError::TransactionGasLimitMoreThanAvailableBlockGas {
+                    transaction_gas_limit: transaction.gas_limit(),
+                    block_available_gas,
+                }
+                .into())
+            }
+
+            EvmConfig::fill_tx_env(evm.tx_mut(), transaction, *sender, ());
+
+            // Execute transaction.
+            let ResultAndState { result, state } = evm.transact().map_err(move |err| {
+                // Ensure hash is calculated for error log, if not already done
+                BlockValidationError::EVM {
+                    hash: transaction.recalculate_hash(),
+                    error: err.into(),
+                }
+            })?;
+            evm.db_mut().commit(state);
```

Comment:
just in mem

tbh I find the entire bundle state thing super confusing and actually not sure if this is need for the single block executor

Reply:
It's so that we can cache effectively across blocks while also keeping all the created/destroyed information in one place (vs clone'ing it / recalculating it like we did before with `PostState`)